p2p: remove tor v2 support

jonatack opened the pull request:
![image](https://user-images.githubusercontent.com/2415484/120018909-4d425a00-bfd7-11eb-83c9-95a3dac97926.jpeg)

This patch removes support in Bitcoin Core for Tor v2 onions, which are already removed from the release of Tor 0.4.6.

- no longer serialize/deserialize and relay Tor v2 addresses
- ignore incoming Tor v2 addresses
- remove Tor v2 addresses from the addrman and peers.dat on node launch
- update generate-seeds.py to ignore Tor v2 addresses
- remove Tor v2 hard-coded seeds

Tested with tor-0.4.6.1-alpha (no v2 support) and 0.4.5.7 (v2 support). With the latest Tor (no v2 support), this removes all the warnings like those reported with current master in https://github.com/bitcoin/bitcoin/issues/21351

```
<bitcoind debug log>
Socks5() connect to […].onion:8333 failed: general failure

<tor log>
Invalid hostname [scrubbed]; rejecting
```

and the addrman no longer has Tor v2 addresses on launching bitcoind.
```rake
$ ./src/bitcoin-cli -addrinfo
{
  "addresses_known": {
    "ipv4": 44483,
    "ipv6": 8467,
    "torv2": 0,
    "torv3": 2296,
    "i2p": 6,
    "total": 55252
  }
}
```
After recompiling back to current master and restarting with either of the two Tor versions (0.4.5.7 or 0.4.6.1), -addrinfo initially returns 0 Tor v2 addresses and then begins finding them again.

Ran nodes on this patch over the past week on mainnet/testnet/signet/regtest after building with DEBUG_ADDRMAN.

Verified that this patch bootstraps an onlynet=onion node from the Tor v3 hardcoded fixed seeds on mainnet and testnet and connects to blocks and v3 onion peers: `rm ~/.bitcoin/testnet3/peers.dat ; ./src/bitcoind -testnet -dnsseed=0 -onlynet=onion`

![Screenshot from 2021-05-28 00-26-17](https://user-images.githubusercontent.com/2415484/119905021-ea02ea00-bf3a-11eb-875f-27ef57640c49.png)

Tested using `addnode`, `getaddednodeinfo`,`addpeeraddress`, `disconnectnode` and `-addrinfo` that a currently valid, connectable Tor v2 peer can no longer be added:

![Screenshot from 2021-05-30 11-32-05](https://user-images.githubusercontent.com/2415484/120099282-29435d80-c12a-11eb-81b6-5084244d7d2a.png)

Thanks to Vasil Dimov, Carl Dong, and Wladimir J. van der Laan for their work on BIP155 and Tor v3 that got us here.

jonatack committed:
p2p: remove pre-addrv2 onions from SerializeV1Array()

jonatack committed:
p2p: remove torv2/ADDR_TORV2_SIZE from SetTor()

jonatack committed:
p2p: drop onions from IsAddrV1Compatible(), no longer relay torv2

jonatack committed:
p2p: remove torv2 from SetNetFromBIP155Network()

jonatack committed:
contrib: remove torv2 seed nodes

jonatack committed:
test: update feature_proxy to torv3

jonatack commented:
A recent IRC discussion for more context. http://www.erisian.com.au/bitcoin-core-dev/log-2021-05-20.html#l-401

theStack commented:
Concept ACK

jonatack committed:
p2p: ignore torv2-in-ipv6 addresses in SetLegacyIPv6()

Rspigler commented:
Concept ACK

duncandean commented:
Concept ACK. (Thank you for your _service_, tor v2)

0xB10C commented:
Concept ACK

laanwj commented:
Concept ACK, adding 22.0 label

jonatack committed:
p2p: remove BIP155Network::TORV2 from GetBIP155Network()

jonatack committed:
contrib: update generate-seeds.py to ignore torv2 addresses

laanwj commented on contrib/seeds/nodes_main.txt:
```diff
@@ -650,518 +650,6 @@
 [2a0f:df00:0:254::46]:8333
 [2c0f:f598:5:1:1001::1]:8333
 [2c0f:fce8:0:400:b7c::1]:8333
-226eupdnaouu4h2v.onion:8333
```
Please also run `cd contrib/seeds && python3 generate-seeds.py . > ../../src/chainparamsseeds.h` to update `chainparamsseeds.h` (in the same commit?).

jonatack replied:
Thanks! Done.

jonatack committed:
p2p: remove torv2 from CNetAddr::ToStringIP()

jonatack committed:
p2p: remove torv2 in SetIP() and ADDR_TORV2_SIZE constant

jonatack committed:
fuzz: rename CNetAddr/CService deserialize targets

as the changes that follow are incompatible with the inputs.

jonatack commented:
This seems to be ready. Updated the PR description. Will propose release documentation in a follow-up.

willcl-ark commented:
Light tested ACK, rebased on top of master @ 123b401e0acf3b26f149711bdf06d7f6eced1968

Used Tor version 0.4.5.7. with manually configured hidden service and a bitcoin.conf with 

```
listenonion=1
onlynet=onion
proxy=127.0.0.1:9050
```

* Node has not picked up any Tor V2 addresses since startup and has a healthy number of V3 addresses in a short period of time:

<img width="1680" alt="Screenshot 2021-05-28 at 20 16 13" src="https://user-images.githubusercontent.com/6606587/120032326-ec744b00-bff1-11eb-8523-3cb887eb3df5.png">

* Tested bootstrapping without peers.dat and that picked up over 1000 V3 peers in just a few minutes.

* Adding a V2 address check back into src/test/net_test.cpp resulted in expected test failure:

`test/net_tests.cpp(614): error: in "net_tests/cnetaddr_unserialize_v2": check addr.IsValid() has failed`

benthecarman commented on contrib/seeds/generate-seeds.py:
```diff
@@ -46,9 +46,7 @@ def name_to_bip155(addr):
     if addr.endswith('.onion'):
         vchAddr = b32decode(addr[0:-6], True)
-        if len(vchAddr) == 10:
-            return (BIP155Network.TORV2, vchAddr)
```
I think an error message specific for v2 not being supported would be better than the defualt error message

laanwj replied:
I guess `makeseeds.py` is still producing these? It should probably ignore Torv2's from the DNS seeder completely.

jonatack replied:
Thanks, updated `generate-seeds.py` to ignore torv2 rather than raise.

jonatack commented:
Changed `generate-seeds.py` to ignore torv2 addresses rather than raise.

<details><summary><code>git diff 5715134 ef34bda</code></summary><p>

```diff
diff --git a/contrib/seeds/generate-seeds.py b/contrib/seeds/generate-seeds.py
index aaccefcf6f..ae97240203 100755
--- a/contrib/seeds/generate-seeds.py
+++ b/contrib/seeds/generate-seeds.py
@@ -47,8 +47,10 @@ def name_to_bip155(addr):
     if addr.endswith('.onion'):
         vchAddr = b32decode(addr[0:-6], True)
         if len(vchAddr) == 35:
-            assert(vchAddr[34] == 3)
+            assert vchAddr[34] == 3
             return (BIP155Network.TORV3, vchAddr[:32])
+        elif len(vchAddr) == 10:
+            return (BIP155Network.TORV2, vchAddr)
         else:
             raise ValueError('Invalid onion %s' % vchAddr)
     elif addr.endswith('.b32.i2p'):
@@ -98,7 +100,10 @@ def parse_spec(s):
 
     host = name_to_bip155(host)
 
-    return host + (port, )
+    if host[0] == BIP155Network.TORV2:
+        return -1  # TORV2 is no longer supported, so we ignore it
+    else:
+        return host + (port, )
 
 def ser_compact_size(l):
     r = b""
@@ -134,6 +139,8 @@ def process_nodes(g, f, structname):
             continue
 
         spec = parse_spec(line)
+        if spec == -1:  # ignore this entry (e.g. no longer supported addresses like TORV2)
+            continue
         blob = bip155_serialize(spec)
         hoststr = ','.join(('0x%02x' % b) for b in blob)
         g.write(f'    {hoststr},\n')
```
</p></details>

benthecarman commented:
utACK ef34bda6c52d7a59be9cd3451bc3536430410e36

jonatack commented:
Tested using `addnode`, `addpeeraddress`, `disconnectnode` and `-addrinfo` that a tor v2 peer can no longer be added. See screenshot for details.

![Screenshot from 2021-05-29 20-34-03](https://user-images.githubusercontent.com/2415484/120081350-ca8bce80-c0ac-11eb-869c-8ca307147501.png)

duncandean commented:
tACK `ef34bda`

* Using Tor v0.4.4.6.
* Node originally had established connections with Tor v2 peers, now does not connect to v2 peers on patch.
* After a short while, with `onlynet=onion`, node is connected to 8+ Tor v3 peers exclusively.

No Tor v2 addresses known
```
{
  "addresses_known": {
    "ipv4": 309,
    "ipv6": 33,
    "torv2": 0,
    "torv3": 3563,
    "i2p": 0,
    "total": 3905
  }
}
```

* Cannot add v2 address or connect to v2 address with `addnode`.

Rspigler commented:
Could we add a more detailed error message for someone who misses or doesn't read the release docs?

Zero-1729 commented:
tACK ef34bda6c52d7a59be9cd3451bc3536430410e36

Using Tor v0.4.5.8 on MacOS 11.3.1 (20E241)

> **Note**: Tor v2 seed (`bk7yp6epnmcllq72.onion`) used is from [this article](https://stopanddecrypt.medium.com/running-bitcoin-lightning-nodes-over-the-tor-network-2021-edition-489180297d5).
> Started `bitcoind` with `-onlynet=onion -listenonion=1 -proxy=127.0.0.1:9050`

#### Before (on master be1c512437f7aa8215d2e05eda5cc8b58c3d3937): 

- Establishes connection with Tor v2 peers.

```sh
$ bitcoin-cli -addrinfo
{
  "addresses_known": {
    "ipv4": 272,
    "ipv6": 0,
    "torv2": 2598,
    "torv3": 345,
    "i2p": 0,
    "total": 3215
  }
}
```
- Successfully adds tor v2 peer `bk7yp6epnmcllq72.onion` with `addpeeraddress`.

#### After (on [torv2-remove-support](https://github.com/jonatack/bitcoin/tree/torv2-remove-support)  ef34bda6c52d7a59be9cd3451bc3536430410e36):

- No longer connects to Tor v2 peers.

```sh
{
  "addresses_known": {
    "ipv4": 271,
    "ipv6": 0,
    "torv2": 0,
    "torv3": 352,
    "i2p": 0,
    "total": 623
  }
}
```

- Establishes manual connection with `bk7yp6epnmcllq72.onion` (Tor v2)  using `addpeeraddress` fails.
- Adding `bk7yp6epnmcllq72.onion` via `addnode` also doesn't work.

<details>
    <summary>Outputs</summary>

    $ bitcoin-cli addpeeraddress bk7yp6epnmcllq72.onion 8333  
    {
        "success": false
    }
    $ bitcoin-cli addnode bk7yp6epnmcllq72.onion onetry   
    $ bitcoin-cli disconnectnode bk7yp6epnmcllq72.onion
    error code: -29
    error message:
    Node not found in connected nodes

</details>

Can verify connections to Tor v2 nodes no longer established, including when using `addnode` & `addpeeraddress` after patch (ef34bda6c52d7a59be9cd3451bc3536430410e36) 👍 .

jonatack commented:
Updated https://github.com/bitcoin/bitcoin/pull/22050#issuecomment-850880216 to also mention `getaddednodeinfo`.

> Could we add a more detailed error message for someone who misses or doesn't read the release docs?

Good question, are you thinking of `addpeeraddress` or `addnode`? The former is a hidden test/debug RPC. The latter can be a valuable anti-Eclipse measure to help ensure a node has at least one honest peer, but it is presumably used by informed users.

I was kind of hoping that updating `doc/tor.md` and adding an announcement toward the top of the release notes (both in a follow-up), along with communication on twitter and the fediverse (and by the Tor Project, who propose to communicate together for the transition) and maybe resources like the ML and Bitcoin Optech would be enough.

This is a one-off event, and very soon the Tor v2 network will no longer be operating at all; according to the Tor AMA event last Friday and the metrics at https://metrics.torproject.org/versions.html, the v2 net is apparently already considerably less robust than the v3 one.

jonatack commented:
Updated https://github.com/bitcoin/bitcoin/pull/22050#issuecomment-850880216 to also mention `getaddednodeinfo`.

> Could we add a more detailed error message for someone who misses or doesn't read the release docs?

Good question. Are you thinking of `addpeeraddress` or `addnode`? The former is a hidden test/debug RPC. The latter can be a valuable anti-Eclipse measure to help ensure a node has at least one honest peer, but it is presumably used by informed users.

I was kind of hoping that updating `doc/tor.md` and adding an announcement toward the top of the release notes (both in a follow-up), along with communication on twitter and the fediverse (and by the Tor Project, who propose to communicate together for the transition) and maybe resources like the ML and Bitcoin Optech would be enough.

This is a one-off event and soon the Tor v2 network will no longer be operating at all; according to the Tor AMA event last Friday, the v2 net is already less robust than the v3 one. This should soon be visible in the metrics at https://metrics.torproject.org/versions.html.

jonatack committed:
p2p, refactor: extract OnionToString() from CNetAddr::ToStringIp()

Rspigler commented:
> I was kind of hoping that updating doc/tor.md and adding an announcement toward the top of the release notes (both in a follow-up), along with communication on twitter and the fediverse (and by the Tor Project, who propose to communicate together for the transition) and maybe resources like the ML and Bitcoin Optech would be enough.

Good point, I think if we do all that, it should be enough

kiminuo commented on contrib/seeds/generate-seeds.py:
```diff
@@ -100,7 +100,10 @@ def parse_spec(s):
 
-    return host + (port, )
+    if host[0] == BIP155Network.TORV2:
+        return -1  # TORV2 is no longer supported, so we ignore it
```
Would it be possibly better to return `None` here?

jonatack replied:
Yes, I asked this in https://github.com/bitcoin/bitcoin/pull/22050#issuecomment-850808537, happy to update if reviewers think it's worth changing.

kiminuo replied:
Ah, sorry. `None` +1 from me :)

jonatack replied:
Ok, I have a couple minor refactorings ready that aren't essential to removing tor v2 support and propose to insert this change into that commit, unless I need to update this patch for another reason.

jonatack replied:
done

kiminuo commented on contrib/seeds/generate-seeds.py:
```diff
@@ -46,11 +46,11 @@ def name_to_bip155(addr):
-            return (BIP155Network.TORV2, vchAddr)
-        elif len(vchAddr) == 35:
-            assert(vchAddr[34] == 3)
+        if len(vchAddr) == 35:
+            assert vchAddr[34] == 3
             return (BIP155Network.TORV3, vchAddr[:32])
+        elif len(vchAddr) == 10:
+            return (BIP155Network.TORV2, vchAddr)
```
This re-ordering change looks unnecessary now. The diff would be smaller when you revert this 4 line change. :-)

jonatack replied:
Most frequent path first.

kiminuo commented on src/netaddress.h:
```diff
@@ -334,31 +330,22 @@ class CNetAddr
                 memcpy(arr, IPV4_IN_IPV6_PREFIX.data(), prefix_size);
                 memcpy(arr + prefix_size, m_addr.data(), m_addr.size());
                 return;
-            case NET_ONION:
```
L301 mentions that `GetBIP155Network()` *Must not be called for IsInternal() objects.* 

Would it make sense to add a note about Tor v2 not being supported anymore? The reason why it may make sense is simply because [BIP155](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki) is written with Tor v2 in mind (and that won't change, or will it?) and the function name is `GetBIP155Network`. So we are about to deviate from BIP155. For a good reason but still.

kiminuo replied:
The idea was to change:

```cpp
/**
 * Get the BIP155 network id of this address.
 * Must not be called for IsInternal() objects.
 * @returns BIP155 network id
 */
BIP155Network GetBIP155Network() const;
```

to something like

```cpp
/**
 * Get the BIP155 network id of this address.
 * Must not be called for IsInternal() objects.
 * @returns BIP155 network id except TORV2 which is no longer supported
 */
BIP155Network GetBIP155Network() const;
```

but it's probably an overkill.

jonatack replied:
SGTM, will add to the follow-up refactoring commit unless this patch needs a more substantial update.

jonatack replied:
Updated
```diff
-         * @returns BIP155 network id
+         * @returns BIP155 network id, except TORV2 which is no longer supported.
```

kiminuo commented:
@jonatack `$ ./src/bitcoin-cli -addrinfo` reports `torv2` entry as [mentioned in OP](https://github.com/bitcoin/bitcoin/pull/22050#issue-651655132). The plan is to remove this later, I guess, right?

jonatack commented:
> @jonatack `$ ./src/bitcoin-cli -addrinfo` reports `torv2` entry as [mentioned in OP](https://github.com/bitcoin/bitcoin/pull/22050#issue-651655132). The plan is to remove this later, I guess, right?

Indeed yes, the reason for -addrinfo returning torv2 and torv3 is for this transition period so the user can see if their node knows enough torv3 peers. When torv2 peers are gone, we can simplify -addrinfo to return only "onion" instead.

reemuru commented:
tACK ef34bda. Run on mainnet, Fedora 34 stable.

```bash
[user@fedora bitcoin]$ tor --version
Jun 01 14:44:30.057 [warn] Tor was compiled with zstd 1.4.9, but is running with zstd 1.5.0. For safety, we'll avoid using advanced zstd functionality.
Tor version 0.4.5.8.
```

<details hidden><summary>before log</summary>

    [user@fedora ~]$ bitcoin/src/bitcoin-cli -addrinfo
    {
      "addresses_known": {
        "ipv4": 39776,
        "ipv6": 7431,
        "torv2": 17,
        "torv3": 8,
        "i2p": 0,
        "total": 47232
      }
    }

</details>

<details hidden><summary>after log</summary>

    [user@fedora ~]$ bitcoin/src/bitcoin-cli -addrinfo
    {
      "addresses_known": {
      "ipv4": 39729,
      "ipv6": 7425,
      "torv2": 0,
      "torv3": 416,
      "i2p": 0,
      "total": 47570
      }
    }

</details>

jonatack commented:
Five ACKs now, thank you everyone for testing and reviewing :purple_heart:

MarcoFalke commented:
cc @laanwj from https://github.com/bitcoin/bitcoin/pull/22050#issuecomment-847937835

laanwj commented on contrib/seeds/generate-seeds.py:
```diff
@@ -136,6 +139,8 @@ def process_nodes(g, f, structname):
             continue
 
         spec = parse_spec(line)
+        if spec == -1:  # ignore this entry (e.g. no longer supported addresses like TORV2)
```
I would prefer using the sentinel value `None` instead of `-1` then check with `spec is None` (just a small nit though, no need to change if no other changes are necessary).

jonatack replied:
done

laanwj commented on src/netaddress.cpp:
```diff
@@ -160,10 +145,6 @@ void CNetAddr::SetLegacyIPv6(Span<const uint8_t> ipv6)
         // IPv4-in-IPv6
         m_net = NET_IPV4;
         skip = sizeof(IPV4_IN_IPV6_PREFIX);
-    } else if (HasPrefix(ipv6, TORV2_IN_IPV6_PREFIX)) {
```
Don't we need some kind of check here to be able to filter out torv2-in-ipv6 in old style `addr` messages?

jonatack replied:
Yes, I thought they would be invalid (so ignored) without the removed code, but it may be better to be explicit about it with something like:
```cpp
    } else if (HasPrefix(ipv6, TORV2_IN_IPV6_PREFIX)) {
        // TORv2-in-IPv6 (unsupported). Unserialize as !IsValid(), thus ignoring them.
        // Mimic a default-constructed CNetAddr object which is !IsValid() and thus
        // will not be gossiped, but continue reading next addresses from the stream.
        m_net = NET_IPV6;
        m_addr.assign(ADDR_IPV6_SIZE, 0x0);
        return;
```

jonatack replied:
done

jonatack commented:
Changed the torv2-in-ipv6 case to deserialize explicitly to an invalid address which will be ignored, changed the ignore value in generate-seeds.py from `-1` to `None`, extracted the onion-to-string code to a helper, and updated a Doxygen comment.

<details><summary><code>git diff ef34bda f924515</code></summary><p>

```diff
diff --git a/contrib/seeds/generate-seeds.py b/contrib/seeds/generate-seeds.py
index ae97240203..0ce868bfcc 100755
--- a/contrib/seeds/generate-seeds.py
+++ b/contrib/seeds/generate-seeds.py
@@ -101,7 +101,7 @@ def parse_spec(s):
     host = name_to_bip155(host)
 
     if host[0] == BIP155Network.TORV2:
-        return -1  # TORV2 is no longer supported, so we ignore it
+        return None  # TORV2 is no longer supported, so we ignore it
     else:
         return host + (port, )
 
@@ -139,7 +139,7 @@ def process_nodes(g, f, structname):
             continue
 
         spec = parse_spec(line)
-        if spec == -1:  # ignore this entry (e.g. no longer supported addresses like TORV2)
+        if spec == None:  # ignore this entry (e.g. no longer supported addresses like TORV2)
             continue
         blob = bip155_serialize(spec)
         hoststr = ','.join(('0x%02x' % b) for b in blob)
diff --git a/src/netaddress.cpp b/src/netaddress.cpp
index e46f9c2568..0ae8dd0698 100644
--- a/src/netaddress.cpp
+++ b/src/netaddress.cpp
@@ -145,6 +145,13 @@ void CNetAddr::SetLegacyIPv6(Span<const uint8_t> ipv6)
         // IPv4-in-IPv6
         m_net = NET_IPV4;
         skip = sizeof(IPV4_IN_IPV6_PREFIX);
+    } else if (HasPrefix(ipv6, TORV2_IN_IPV6_PREFIX)) {
+        // TORv2-in-IPv6 (unsupported). Unserialize as !IsValid(), thus ignoring them.
+        // Mimic a default-constructed CNetAddr object which is !IsValid() and thus
+        // will not be gossiped, but continue reading next addresses from the stream.
+        m_net = NET_IPV6;
+        m_addr.assign(ADDR_IPV6_SIZE, 0x0);
+        return;
     } else if (HasPrefix(ipv6, INTERNAL_IN_IPV6_PREFIX)) {
         // Internal-in-IPv6
         m_net = NET_INTERNAL;
@@ -587,31 +594,26 @@ static std::string IPv6ToString(Span<const uint8_t> a, uint32_t scope_id)
     return r;
 }
 
+static std::string OnionToString(const Span<const uint8_t>& addr)
+{
+    uint8_t checksum[torv3::CHECKSUM_LEN];
+    torv3::Checksum(addr, checksum);
+    // TORv3 onion_address = base32(PUBKEY | CHECKSUM | VERSION) + ".onion"
+    prevector<torv3::TOTAL_LEN, uint8_t> address{addr.begin(), addr.end()};
+    address.insert(address.end(), checksum, checksum + torv3::CHECKSUM_LEN);
+    address.insert(address.end(), torv3::VERSION, torv3::VERSION + sizeof(torv3::VERSION));
+    return EncodeBase32(address) + ".onion";
+}
+
 std::string CNetAddr::ToStringIP() const
 {
     switch (m_net) {
     case NET_IPV4:
         return IPv4ToString(m_addr);
-    case NET_IPV6: {
+    case NET_IPV6:
         return IPv6ToString(m_addr, m_scope_id);
-    }
     case NET_ONION:
-        switch (m_addr.size()) {
-        case ADDR_TORV3_SIZE: {
-
-            uint8_t checksum[torv3::CHECKSUM_LEN];
-            torv3::Checksum(m_addr, checksum);
-
-            // TORv3 onion_address = base32(PUBKEY | CHECKSUM | VERSION) + ".onion"
-            prevector<torv3::TOTAL_LEN, uint8_t> address{m_addr.begin(), m_addr.end()};
-            address.insert(address.end(), checksum, checksum + torv3::CHECKSUM_LEN);
-            address.insert(address.end(), torv3::VERSION, torv3::VERSION + sizeof(torv3::VERSION));
-
-            return EncodeBase32(address) + ".onion";
-        }
-        default:
-            assert(false);
-        }
+        return OnionToString(m_addr);
     case NET_I2P:
         return EncodeBase32(m_addr, false /* don't pad with = */) + ".b32.i2p";
     case NET_CJDNS:
diff --git a/src/netaddress.h b/src/netaddress.h
index d539bd51ff..0d04ab88fb 100644
--- a/src/netaddress.h
+++ b/src/netaddress.h
@@ -299,7 +299,7 @@ class CNetAddr
         /**
          * Get the BIP155 network id of this address.
          * Must not be called for IsInternal() objects.
-         * @returns BIP155 network id
+         * @returns BIP155 network id, except TORV2 which is no longer supported.
          */
         BIP155Network GetBIP155Network() const;
```
</p></details>

laanwj commented:
Changes look good to me now.
Code review ACK 5d82a57db4f67506a4e80d186ba76f3a8665e147
Tested ACK of an earlier commit, which I tested for a while on one of my nodes. I noticed no difference in behavior except for no longer connecting to Tor v2 nodes.

vasild reviewed (commented):
ACK 5d82a57db4f67506a4e80d186ba76f3a8665e147

Some followups in https://github.com/bitcoin/bitcoin/pull/22179.

jonatack commented on src/netaddress.cpp:
```diff
@@ -72,14 +65,6 @@ bool CNetAddr::SetNetFromBIP155Network(uint8_t possible_bip155_net, size_t addre
         throw std::ios_base::failure(
             strprintf("BIP155 IPv6 address with length %u (should be %u)", address_size,
                       ADDR_IPV6_SIZE));
-    case BIP155Network::TORV2:
-        if (address_size == ADDR_TORV2_SIZE) {
-            m_net = NET_ONION;
-            return true;
-        }
-        throw std::ios_base::failure(
-            strprintf("BIP155 TORv2 address with length %u (should be %u)", address_size,
-                      ADDR_TORV2_SIZE));
```
Thanks, I'll have a look.
